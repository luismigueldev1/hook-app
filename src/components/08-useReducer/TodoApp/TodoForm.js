import React from "react";
import PropTypes from "prop-types";
import { useForm } from "../../../hooks/useForm";

function TodoForm({ handleAddTodo }) {
  const [values, handleInputChange, reset] = useForm({
    task: "",
  });

  const handleSubmit = (event) => {
    event.preventDefault();

    if (values.task.trim().length <= 1) {
      return;
    }

    const newTodo = {
      id: new Date().getTime(),
      task: values.task,
      done: false,
    };
    handleAddTodo(newTodo);
    reset();
  };

  return (
    <>
      <h4>Agregar TODO</h4>
      <hr />

      <form onSubmit={handleSubmit}>
        <input
          className="form-control"
          type="text"
          name="task"
          placeholder="New task"
          autoComplete="off"
          value={values.task}
          onChange={handleInputChange}
        />
        <button
          type="submit"
          className="btn btn-outline-primary  btn-block mt-2"
        >
          Add New Task
        </button>
      </form>
    </>
  );
}

TodoForm.propTypes = {
  handleAddTodo: PropTypes.func.isRequired,
};

export default TodoForm;
