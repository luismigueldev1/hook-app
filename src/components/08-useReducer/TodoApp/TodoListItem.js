import React from "react";
import PropTypes from "prop-types";

function TodoListItem({ todo, handleDelete, handleToogle }) {
  return (
    <li className="list-group-item d-flex align-items-center justify-content-between">
      <p className="m-0">{todo.done ? <del>{todo.task}</del> : todo.task}</p>
      <div className="buttons">
        <button
          className={`btn btn-sm  ${
            todo.done ? " btn-warning" : " btn-primary"
          }`}
          onClick={() => handleToogle(todo.id)}
        >
          {todo.done}
          {todo.done ? "Pending" : "Done"}
        </button>

        <button
          className="btn btn-sm btn-danger ml-2"
          onClick={() => handleDelete(todo.id)}
        >
          Delete
        </button>
      </div>
    </li>
  );
}

TodoListItem.propTypes = {
  todo: PropTypes.object.isRequired,
  handleDelete: PropTypes.func.isRequired,
  handleToogle: PropTypes.func.isRequired,
};

export default TodoListItem;
