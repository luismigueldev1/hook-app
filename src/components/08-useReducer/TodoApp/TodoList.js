import React from "react";
import TodoListItem from "./TodoListItem";
import PropTypes from "prop-types";

function TodoList({ todos, handleDelete, handleToogle }) {
  return (
    <ul className="p-0 list-group ">
      {todos.map((todo) => (
        <TodoListItem
          key={todo.id}
          todo={todo}
          handleDelete={handleDelete}
          handleToogle={handleToogle}
        />
      ))}
    </ul>
  );
}

TodoList.propTypes = {
  todos: PropTypes.array.isRequired,
  handleDelete: PropTypes.func.isRequired,
  handleToogle: PropTypes.func.isRequired,
};

export default TodoList;
