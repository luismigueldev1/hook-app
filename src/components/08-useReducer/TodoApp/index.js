import React, { useEffect, useReducer } from "react";

import { todoReducer } from "./todoReducer";
import TodoList from "./TodoList";
import TodoForm from "./TodoForm";

const init = () => {
  return JSON.parse(localStorage.getItem("todos")) || [];
};

export default function TodoApp() {
  const [todos, dispatch] = useReducer(todoReducer, [], init);

  useEffect(() => {
    localStorage.setItem("todos", JSON.stringify(todos));
  }, [todos]);

  const handleDelete = (id) => {
    const action = {
      type: "DELETE",
      payload: id,
    };
    dispatch(action);
  };

  const handleToogle = (id) => {
    dispatch({
      type: "TOOGLE",
      payload: id,
    });
  };

  const handleAddTodo = (newTodo) => {
    dispatch({
      type: "ADD",
      payload: newTodo,
    });
  };

  return (
    <div>
      <h1>
        Todo App<small>({todos.length})</small>
      </h1>
      <hr />

      <div className="row">
        <div className="col-sm-12 col-md-7 mb-2">
          <TodoList
            todos={todos}
            handleDelete={handleDelete}
            handleToogle={handleToogle}
          />
        </div>

        <div className="col">
          <TodoForm handleAddTodo={handleAddTodo} />
        </div>
      </div>
    </div>
  );
}
