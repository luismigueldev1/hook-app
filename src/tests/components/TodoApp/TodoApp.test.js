import { act } from "@testing-library/react";
import { mount, shallow } from "enzyme";
import TodoApp from "../../../components/08-useReducer/TodoApp/index";
import { demoTodos } from "../../fixtures/demoTodos";

describe("Pruebas en <TodoApp/>", () => {
  const wrapper = shallow(<TodoApp />);

  Storage.prototype.setItem = jest.fn(() => {});

  test("debe renderizar correctamente", () => {
    expect(wrapper).toMatchSnapshot();
  });

  //   test("debe de agregar un TODO", () => {
  //     const wrapper = mount(<TodoApp />);
  //     console.log(wrapper);

  //     act(() => {
  //       wrapper.find("TodoForm").prop("handleAddTodo")(demoTodos[0]);
  //       wrapper.find("TodoForm").prop("handleAddTodo")(demoTodos[1]);
  //     });

  //     expect(wrapper.find("h1").text().trim()).toBe("Todo App (2)");
  //     expect(localStorage.setItem).toHaveBeenCalledTimes(2);
  //   });
});
