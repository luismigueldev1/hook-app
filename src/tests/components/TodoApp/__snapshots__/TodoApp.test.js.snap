// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoApp/> debe renderizar correctamente 1`] = `
<div>
  <h1>
    Todo App
    <small>
      (
      0
      )
    </small>
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-sm-12 col-md-7 mb-2"
    >
      <TodoList
        handleDelete={[Function]}
        handleToogle={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col"
    >
      <TodoForm
        handleAddTodo={[Function]}
      />
    </div>
  </div>
</div>
`;
